waspc/e2e-test/Util.hs

```diff
@@ -0,0 +1,96 @@
+      -- Overwrite default Wasp project .gitignore to exclude node_modules.
+      -- This is to ensure our golden test dirs won't commit it.
+      "cp " ++ ("overlays" </> ".gitignore") ++ " " ++ testProjectName
+    ]
+
+waspCliNewAndCompile :: String -> ShellCommand
+waspCliNewAndCompile testProjectName =
+  intercalate " && " [waspCliNew testProjectName, "cd " ++ testProjectName, "wasp-cli compile"]
+
+waspCliNewAndCompileAndMigrate :: FilePath -> FilePath -> String -> ShellCommand
+waspCliNewAndCompileAndMigrate _cwd overlayFp testProjectName =
+  let generatedMigrationsDir = joinPath [".wasp", "out", "db", "migrations"]
+      waspMigrationsDir = "migrations"
+      migrationName = "foo"
+   in intercalate
+        " && "
+        [ waspCliNewAndCompile testProjectName,
+          -- Append content from the overlay file (e.g. an entity) to the end of the wasp file
+          "cat " ++ overlayFp ++ " >> main.wasp",
+          -- Migrate using a migration name to avoid Prisma asking via CLI
+          "wasp-cli db migrate-dev " ++ migrationName,
+          -- Rename both migrations to remove the date-specific portion of the directory to something static
+          "mv " ++ (waspMigrationsDir </> ("*" ++ migrationName)) ++ " " ++ (waspMigrationsDir </> ("no-date-" ++ migrationName)),
+          "mv " ++ (generatedMigrationsDir </> ("*" ++ migrationName)) ++ " " ++ (generatedMigrationsDir </> ("no-date-" ++ migrationName))
+        ]
+
+-- | This runs a golden test by creating a Wasp project, running commands,
+-- and then comparing the outputs to the golden directory.
+runGoldenTest :: FilePath -> String -> (String -> String) -> IO TestTree
+runGoldenTest cwd goldenDirName command = do
+  let testProjectName = goldenDirName ++ "Test"
+
+  -- TODO: Save stdout/error as log file for "contains" checks
+  let testProjectDirAbsFp = cwd </> "e2e-test" </> "goldens" </> testProjectName
+
+  -- Remove existing Wasp projects from prior test runs
+  callCommand $ "rm -rf " ++ testProjectDirAbsFp
+
+  -- Run the series of commands within the context of a given project
+  callCommand $ "cd " ++ (cwd </> "e2e-test" </> "goldens") ++ " && " ++ command testProjectName
+
+  testProjectFps <- listAllFilesFromTestProjectDir testProjectDirAbsFp
+  manifestFp <- writeManifest cwd testProjectFps testProjectDirAbsFp
+  return $
+    testGroup
+      goldenDirName
+      [ goldenVsFileDiff
+          testProjectFp -- test name
+          (\ref new -> ["diff", "-u", ref, new])
+          goldenFp
+          testProjectFp
+          (return ()) -- a no-op command that normally generates the file under test, but we did that in bulk above
+        | testProjectFp <- manifestFp : testProjectFps,
+          let goldenFp = unpack $ replace (pack testProjectName) (pack goldenDirName) (pack testProjectFp) -- TODO: clean up after SP transition
+      ]
+  where
+    -- TODO: Ideally we would not ignore `package.json`, but in CI the order of dependencies differs :/
```

Comment:
Interesting! We should sort them before generation then. Can you create an issue?

Reply:
This should be the Node package.json output from `npm install`, not your json file... can we control how Node outputs? Or do you mean we should be somehow parsing and rewriting that file during your stuff?